[SNAT] add "need to frag" ICMP support

sahid opened the pull request:
Assuming that, for a given endpoint sending a packet to the world of a size bigger than the accepted MTU in the networking path.

A networking equipment in this networking path would return an ICMP Error Destination Unreachable with code `ICMP_FRAG_NEEDED`.

According to the RFE 5508 [0]. The networking equipment should embed the original packet in its response.

```
Destination Unreachable Message

    0                   1                   2                   3
    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |     Type      |     Code      |          Checksum             |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                             unused                            |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |      Internet Header + 64 bits of Original Data Datagram      |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

Previously, because of the missing support of `ICMP_FRAG_NEEDED`, the NAT session to transfert the reply packet to the endpoint was never resolved since the source IP addr of the reply packet would be IP addr of the networking equipment.

```
Pod-X      : 10.244.0.236
Cilium-Nat : 192.168.49.2
Router-Y   : 192.168.49.1
Host-Y     : 10.1.0.100

                                         
                                  +------------+
                                  | Pod-X      |
                                  +------------+
            Cilium Host                  |
            --------------------+--------+-------------------
                                |  | ICMP Packet Pod-X to Host-y
                                |  |   10.244.0.256/10.1.0.100
                                |  v
                          +--------------+
                          | Cilium-Nat   |
                          +--------------+
                                |  | ICMP Packet Cilium-Nat to Host-y
                                |  |   192.168.49.2/10.1.0.100 
    World                       |  v
   ----------------+------------+----------------
                   |  ^
                   |  | ICMP Error Router-Y to Cilium-Nat
                   |  |   192.168.49.1/192.168.49.2
                   |  |      with in Payload ICMP 192.168.49.2/10.1.0.100
            +-------------+
            | Router-Y    |
            +-------------+
                   |
   ----------------+-------+--------
                           |
                         Host-Y
```

Mainly implementation consists:

- On adding to `snat_v{4,6}_rev_nat()` the support for reading the original packet that has been embedded in the ICMP Error to extract headers needed to resolve rev-NAT session. For `snat_v{4,6}_nat()` it is considered to drop the packet.
- It will be also mandatory to rewrite embedded packets to NAT original headers because the source IP addr of the original header is the backend node.

Finally reply packet received by endpoint will be:

```
                   |  | ICMP Error Router-Y to Pod-X
                   |  |   192.168.49.1/10.244.0.256
                   |  |      with in Payload ICMP 10.244.0.256/10.1.0.100
```

Fixes: #12968

[0] https://www.rfc-editor.org/rfc/rfc5508.html#section-7.1

sahid committed:
bpf/nat: implement support of ICMP6 Error packet too big

Running in masquerading BPF mode.

Assuming that, Pod is sending a packet to the world of a size bigger
than the accepted MTU in the networking path.

A networking equipment in that networking path would return an ICMP
Error packet with code PACKET_TOO_BIG. According to the RFE 5508,
that networking equipment should embed the original packet in its
response.

Previously because of the missing support of PACKET_TOO_BIG, the NAT
session to return the packet to Pod was never resolved. Meaning that
the ICMP Error packet got dropped.

 a) The src IP of the packet becomes the networking equipment that is
 generating the error response.  b) The protocol used for the response
 is ICMP where the packet sent may be TDP/UPD.

This change is updating the NAT process to read the embedded packet in
the ICMP Error packet that to determine the NAT session and send back
the packet to Pod.

This change is for IPv6.

Fixes: #12968
Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

kkourt commented on bpf/lib/nat.h:
```diff
@@ -552,6 +552,41 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
+
+					// RFC 5508 indicates that original
+					// IPV4+ICMP headers are added in
+					// payload.
```
I'd like to check that my understanding is correct.

We try to deal with the following case (#12968)
```
Pod A (MTU 1500) -> Node A (MTU 1500) -> Switch X (MTU 1420) -> Node B (MTU 1500)
```
Specifically, what to do when we get a `ICMP_FRAG_NEEDED` error from, e.g., Switch X, above.


Quoting RFC5508:
```
ICMP Error Messages - ICMP Error messages provide signaling for IP.
     All ICMP Error messages are characterized by the fact that they
     embed the original datagram that triggered the ICMP Error message.
     The original datagram embedded within the ICMP Error payload is
     also referred to as the "Embedded packet" throughout the document.
     Unlike ICMP Query messages, ICMP Error messages do not have a Query
     Id in the ICMP header.
```
Wouldn't be possible for the embedded packet to be any other packet? (e.g., TCP or UDP)? 
We seem to be assuming embedded packet will always be an ICMP_ECHO. Is there a reason for this (complexity, presumably)?

sahid replied:
Yes you are totally right I was making wrong assumption saying that the embedded packet will always be an ICMP_ECHO (I guess for IPPROTO ICMP, it can't be anything else, right?)

I'm currently fixing that part, I should provide new PR shortly.

kkourt commented on bpf/tests/prog_test/prog_test.go:
```diff
@@ -341,6 +341,63 @@ func testCt4Rst(spec *ebpf.Collection) error {
+	)
+
+	bpfRet, _, err := prog.Test(packetIn.Bytes())
+	if err != nil {
```
I believe the second argument that is omitted here is the output packet?
Are there any specific expectations that we have for the output packet? would it make sense to test them?

sahid replied:
I didn't think about testing that one I was more focus on ensuring behavior of the NAT process. I will have a look and let you know.

kkourt reviewed (changes requested):
Hi!

Thanks again for the PR!

Not sure I fully understand what the code is trying to do.

Could you please add additional text in the commit messages to explain what the intention of the patch is? It's OK to repeat the PR text.

sahid commented on bpf/lib/nat.h:
```diff
@@ -999,6 +1062,7 @@ snat_v6_process(struct __ctx_buff *ctx, enum nat_dir dir,
 		return DROP_INVALID;
 
 	nexthdr = ip6->nexthdr;
+
```
unnecessary change I will remove that for my next PR. waiting for more comments first.

sahid commented on bpf/lib/nat.h:
```diff
@@ -1007,6 +1071,7 @@ snat_v6_process(struct __ctx_buff *ctx, enum nat_dir dir,
 	ipv6_addr_copy(&tuple.daddr, (union v6addr *)&ip6->daddr);
 	ipv6_addr_copy(&tuple.saddr, (union v6addr *)&ip6->saddr);
 	tuple.flags = dir;
+
```
unnecessary change I will remove that for my next PR. waiting for more comments first.

sahid commented on bpf/lib/nat.h:
```diff
@@ -541,16 +541,79 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
+					tuple.daddr = iphdr.saddr;
+
+					off += ipv4_hdrlen(&iphdr);
+					switch (iphdr.protocol) {
```
nit: /iphdr.protocol/tuple.nexthdr

sahid commented on bpf/Makefile:
```diff
@@ -312,9 +312,9 @@ tests/bpf_nat_tests.o: tests/bpf_nat_tests.ll
-go_prog_test: $(BPF_TEST)
-	$(GO) test -tags privileged_tests -exec sudo ./tests/prog_test
+.PHONY: go_prog_ct_test
+go_prog_ct_test: $(BPF_TEST)
```
I missed to update `.github/workflows/lint-bpf-checks.yaml`. I will fix that with my next PR.

pchaigno commented on bpf/lib/nat.h:
```diff
@@ -552,6 +554,60 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
+
+				off += sizeof(icmphdr);
+				if (ctx_load_bytes(ctx, off, &iphdr,
+						   sizeof(iphdr)) < 0)
```
On XDP, these `ctx_load_bytes` are not converted into a helper call [but using `memcpy`](https://github.com/cilium/cilium/blob/82c742f2e9fb65fb3fc392c7dcb3b4e22b69c650/bpf/include/bpf/ctx/xdp.h#L34). That can be quite expensive in the number of instructions used. Given we only need four fields from the IP header, maybe we could simply read those four fields? (See for instance how we load the L4 fields we need below.)

pchaigno replied:
@sahid Answering here, so we can close the thread once resolved. 

> The saddr and daddr are at the very end, we will still have to load the full object, did I missed something?

If you only want to load e.g. the `saddr` and `daddr`, you can change `off` to start reading for the offset where `saddr` is and change `sizeof(iphdr)` to read only 8 bytes (2 IPv4 addresses). No need to read `sizeof(iphdr)` bytes.

sahid replied:
Oh right! I think I see where you want me to go now. Thank!

sahid replied:
I have made change for IPv4 only, could you let me know your thinking about it when you have a moment?

pchaigno replied:
Looks good but I wouldn't put it all inside one if condition and we also shouldn't hardcode the sizes.

sahid replied:
ACK I will fix that and work on IPv6. Thanks.

sahid replied:
I fixed you point and addressed IPv6. 

~~BTW and I should probably have asked this question before, since we are now calling `ctx_load_bytes` more times, we are also going to call `memcpy` more times, that is not going to increase the number of instructions?~~ I guess if we reduce the size to copy we will reduce the number of interactions with registers.

pchaigno replied:
Yep, I'd expect the number of instructions for the three `memcpy`s to be less than the number of instructions for the single larger `memcpy`. Note however that, on the tc side, `ctx_load_bytes` is converted into helper calls, so this change will lead to more instructions for tc programs. Hopefully, we're not just trading a program size issue on XDP for one on tc :crossed_fingers:

sahid replied:
Looks like we are still exceeding the limit :/

```
	 libbpf: load bpf program failed: Argument list too long
	 libbpf: Program too large (4264 insns), at most 4096 insns
```

pchaigno replied:
Yep, it's even worse because net-next is now failing with a complexity issue in addition to the program size issue :roll_eyes: 

I've asked internally our NAT experts to take a look at the pull request. There's no need to invest a lot of time in solving those verifier issues if the pull request is not taking the right approach anyway. They may also have other ideas to work around these issues.

sahid replied:
ACK thank you Paul.

sahid replied:
Don't you think that I should revert back the change as it was before?

pchaigno commented on bpf/lib/nat.h:
```diff
@@ -1092,6 +1094,58 @@ snat_v6_process(struct __ctx_buff *ctx, enum nat_dir dir,
+			 * sequence in its headers.
+			 */
+			off += sizeof(icmp6hdr) - sizeof(icmp6hdr.icmp6_dataun.u_echo);
+			if (ctx_load_bytes(ctx, off, &iphdr, sizeof(iphdr)) < 0)
```
Same comment here as for the IPv4 path.

sahid replied:
done

pchaigno reviewed (commented):
There's maybe one change we can make to reduce the number of instructions at the XDP hook point (see below). Not sure it will be enough but probably worth trying.

sahid commented on bpf/lib/nodeport.h:
```diff
@@ -637,6 +637,7 @@ int tail_nodeport_nat_ipv6(struct __ctx_buff *ctx)
 		}
 	}
 #endif
+#define DISABLE_ICMP_SUPPORT
```
Should have defined this before to include nat.h ... I should take some holidays I guess.

sahid commented on bpf/lib/nat.h:
```diff
@@ -507,20 +507,23 @@ static __always_inline __maybe_unused int
 	bool icmp_echoreply = false;
 	__u64 off;
 	int ret;
+#ifndef DISABLE_ICMP_SUPPORT
```
There is a typo for snat snat_v4_process, the right name of this is DISABLE_SNAT_ICMP_SUPPORT. The CI passed because the issue was only for ipv6. I will fix that witth my next change.

brb commented on bpf/lib/nat.h:
```diff
@@ -538,21 +543,80 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
+			break;
+		case ICMP_DEST_UNREACH:
+			switch (icmphdr.code) {
+			case ICMP_FRAG_NEEDED:
```
Do you have a pcap of such a packet? Asking, as I'm wondering whether not only the original packet needs to be rev-SNAT-ed. The top IPv4 hdr might need to be translated too in order for it to reach the sender pod.

sahid replied:
Is that what you are looking for Martynas [0]?

My understanding is that, the top IPv4 hdr has tuple ROUTER, GATEWAY, and it won't help to resolve the NAT session, let me know your thinking. Thanks.

[0]https://github.com/cilium/cilium/pull/18414#issuecomment-1047107255

brb replied:
Thanks.

> From POD point of view:
> ...
> In f6:8e:6d:a1:ac:bb ethertype IPv4 (0x0800), length 592: 192.168.49.1 > 10.244.0.236: ICMP 10.1.0.100 unreachable - need to frag (mtu 800), length 556

Who did the rev-SNAT translation for this packet? I.e., who changed from 192.168.49.2 to 10.244.0.236 in the top-level IPv4 hdr dst addr field? My understanding that your proposed change is changing only the src addr of the internal (= the original packet).

sahid commented on bpf/lib/nat.h:
```diff
@@ -538,21 +543,80 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
+				 */
+				tuple.nexthdr = iphdr.protocol;
+				tuple.saddr = iphdr.daddr;
+				tuple.daddr = iphdr.saddr;
```
@brb The change is done here.

```
From original IP hdr:
--
tuple.daddr is 192.168.49.1   # HOST
tuple.saddr is 192.168.49.2   # ROUTER

Embeded packet IP hdr:
--
iphdr.saddr is 192.168.49.1 # HOST
iphdr.daddr is 10.1.0.100     # DST

Changes
--
tuple.saddr is 10.1.0.100     # DST
tuple.daddr is 192.168.49.1 # HOST

```

1. Here [0] we can see that we create a nat entry with `to_daddr` pointing to the  POD.
2. Here [1] we can see that we create a rtuple entry with saddr pointing to DST and daddr pointing to HOST



[0] https://github.com/cilium/cilium/blob/master/bpf/lib/nat.h#L218
[1] https://github.com/cilium/cilium/blob/master/bpf/lib/nat.h#L229

You probably know all of that but since i'm not sure to understand your whole question I'm basically trying to share with you my understanding.

brb replied:
Could you please attach a pcap with the ICMP NEED FRAG. I'd help me with reviewing the changes.

sahid replied:
Sure give me a moment to build env and share this with you. Thank you.

sahid replied:
@brb Morning,

Here are the pcap from pod and bridge POV.

https://drive.google.com/drive/folders/1BFrLFkXUlyy_sUZ6ryElcI9Il2p-Zhnf?usp=sharing

Let me know if you need anything else. Thanks

brb replied:
Thanks for the pcap! Was it obtained when running with your changes? I'm wondering who did the translation for the marked IP addr.
![need-frag](https://user-images.githubusercontent.com/122571/158400300-7a06a823-74c4-4bdc-b351-d8c29dd7ec77.png)

sahid replied:
Hello Martynas, 

Thank you to have looked at it so fast :-)

My understanding is that. During egress we store a rtuple DST IP/HOST IP with a nat entry pointing to the POD IP. during ingress we enter in the function `snat_v4_rewrite_ingress` [0] to rewrite the packet destination.

[0] https://github.com/cilium/cilium/blob/master/bpf/lib/nat.h#L422

PS: Yes the pcap are obtained with my changes. Without it the packet get dropped.

sahid replied:
~~There is effectively something that is not clear, in the embedded packet the source IP is the router so it seems that we endup with a tuple ROUTER IP/POD IP, the snat does not have such entry in its table...~~ I need more coffee source is the HOST IP :)

sahid replied:
```
Pod-X      : 10.244.0.236
Cilium-Nat : 192.168.49.2
Router-Y   : 192.168.49.1
Host-Y     : 10.1.0.100

                                         
                                  +------------+
                                  | Pod-X      |
                                  +------------+
            Cilium Host                  |
            --------------------+--------+-------------------
                                |  | ICMP Packet Pod-X to Host-y
                                |  |   10.244.0.256/10.1.0.100
                                |  v
                          +--------------+
                          | Cilium-Nat   |
                          +--------------+
                                |  | ICMP Packet Cilium-Nat to Host-y
                                |  |   192.168.49.2/10.1.0.100 
    World                       |  v
   ----------------+------------+----------------
                   |  ^
                   |  | ICMP Error Router-Y to Cilium-Nat
                   |  |   192.168.49.1/192.168.49.2
                   |  |      with in Payload ICMP 192.168.49.2/10.1.0.100
            +-------------+
            | Router-Y    |
            +-------------+
                   |
   ----------------+-------+--------
                           |
                         Host-Y
```

Hello Martynas, I have provided some more details about the setup and scenario. I hope that will help. Let me know if you need any more information. Thanks.

brb replied:
Thanks for the explanation. One question though - is it OK that the embedded original payload (`Payload ICMP 192.168.49.2/10.1.0.100`) is not rev-SNAT-ed from 192.168.49.2 to 10.244.0.236? What does the RFC say about it?

sahid replied:
Nothing is mentioning that the payload should be updated. We should even not drop the packet if the embedded one has an invalid checksum [0]. At least from my understanding.

We are doing end-to-end tests with real use-case using IPSec. I will let you know the result when it's done.

https://www.rfc-editor.org/rfc/rfc5508.html#section-4.1

brb replied:
Could you check what netfilter's SNAT does? You could disable Cilium's BPF masq by setting `--enable-bpf-masquerade=false`.

sahid replied:
I should have think about that before :-) thank you I will do that and let you know.

sahid replied:
Hello Martynas, It seems that the source IP of the  embbeded packet it also translated, meaning the checksum of the ip header is translated as-well. It's quite a big thing to implement this feature. I will add this and get back to you shortly.

  https://drive.google.com/file/d/1ZpGLNOMG6P4-UbOmDS2mdRv2manO-Pun/view

brb replied:
Hey @sahid , I think I am still confused by the output seen in https://github.com/cilium/cilium/pull/18414/files#r827040230. Your change is supposed to rev-SNAT the embedded packet in the ICMP FRAG NEEDED. But in the output we can see that the ICMP FRAG NEEDED is rev-SNAT-ed instead :thinking: 

Another thing is that the required changes might explode the complexity. I'm taking to my this week's TODO to check whether the refactoring snat_v{4,6}_process => do_snat_v{4,6}/do_rev_snat_v{4,6} is feasible. This could help with your PR.

sahid replied:
> Hey @sahid , I think I am still confused by the output seen in https://github.com/cilium/cilium/pull/18414/files#r827040230. Your change is supposed to rev-SNAT the embedded packet in the ICMP FRAG NEEDED. But in the output we can see that the ICMP FRAG NEEDED is rev-SNAT-ed instead thinking

Hum... no no the change is well supposed to rev-SNAT the ICMP FRAG NEEDED. I think we are good for this part. Are you considering that we should not?

The missing part is to replace the source IP of the embedded packet from HOST to POD.

- This is an example with iptables: https://imgur.com/arDHFlQ

> Another thing is that the required changes might explode the complexity. I'm taking to my this week's TODO to check whether the refactoring snat_v{4,6}_process => do_snat_v{4,6}/do_rev_snat_v{4,6} is feasible. This could help with your PR.

Yes, I'm also a bit scary for that as we will have to rewrite two new fields for the embedded packet, the source IP and checksum.
I wanted to work on it after this change, but if you have desires for that I will let you take the lead :-)

brb replied:
> think we are good for this part. Are you considering that we should not?

Re-read the changes. I was mistakenly thinking, that inside the FRAG NEEDED branch you were updating `off`. All good.

> I wanted to work on it after this change, but if you have desires for that I will let you take the lead :-)

Sure, you can work on it. I just want to validate whether the proposed refactoring makes sense.

sahid replied:
Hello @brb the nat part for the embedded packet is ready, ipv4 only. When you have moment to make review and share your thinking., I understand the patch is quite big now and may be painful to review. Thanks in advance.

joestringer replied:
I don't see followup here since April. If I follow right, the question is whether the code is generating correct packets and the hope was to see pcaps to demonstrate this? Could we add tests into the tree that prove this instead?

sahid replied:
Hum I just noticed your comment. I have shared the iptables example demonstrating that the embedded packet get NATed as-well. Also previously there was tests to demonstrate that the code was doing what is expected. But As mentioned in a different threads I had to remove all of them because Cilium proposed a new test framework.

All good now I guess. I have re-implemented tests using the new test framework. Unfortunately for IPv6 there are some issues. I hope that will not be a blocker. Thanks.

sahid commented on bpf/tests/prog_nat_test/prog_test.go:
```diff
@@ -0,0 +1,573 @@
+	return nil
+}
+
+func assertICMPErrorOutputPacketv4(packetOut []byte) error {
```
This is missing assertions for the embedded packet

brb commented on bpf/lib/nat.h:
```diff
@@ -571,7 +639,7 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
 	return dir == NAT_DIR_EGRESS ?
 	       snat_v4_rewrite_egress(ctx, &tuple, state, off, ipv4_has_l4_header(ip4)) :
-	       snat_v4_rewrite_ingress(ctx, &tuple, state, off);
+	       snat_v4_rewrite_ingress(ctx, &tuple, state, off, nexthdr || tuple.nexthdr);
```
`nexthdr || tuple.nexthdr` returns bool, so only `1` or `0` are possible values. The `snat_v4_rewrite_ingress()` function expects it to be a next header though.

sahid replied:
Oh right! thank you Martynas. I fixed it.

brb reviewed (dismissed):
Reviewed the v4 path.

kkourt commented on bpf/lib/nat.h:
```diff
@@ -382,12 +382,82 @@ static __always_inline int snat_v4_rewrite_egress(struct __ctx_buff *ctx,
+	if (state->to_dport != tuple->dport) {
+		switch (nexthdr) {
+		case IPPROTO_TCP:
+		case IPPROTO_UDP:
```
@sahid Thanks for your work on this!

I'll admit that I've lost track of this PR, but it's not clear to me how the changes here relate to handling`ICMP_FRAG_NEEDED`. Am I missing something?

Might be worth adding a comment and/or some text in the commit message to describe what these changes are and why  they are needed.

FWIW, I would suggest to not do any code changes yet. Wait for feedback from other reviewers first, since CI is now green.

sahid replied:
Thanks a lot @kkourt to have took time looking at this IPv4 implementation I have a the IPv6 ready based on that one.

About your ask I have created a method `rewrite_embedded` which changes the source IP of the embedded packet and also the for TCP/UDP the port if it has been changed during the egrees NAT, for ICMP the ID.

The method is quite similar to `rewrite_ingress`, so the diff completely mixed-up the thing which is making the review very difficult.I will try in my next update to move the method, so I would expect the diff to avoid that mix. 

I will also add more comments. That makes lot of sense yes.

As I mentioned I have the IPv6 ready but I will wait to get more comments, probably from @brb :-)

kkourt reviewed (approved):
The PR looks good from my side. I'm not able to follow all the code paths, so I'm hoping @brb will have a closer look.

One minor comment is that it would be great if we could add some more context in:
https://github.com/cilium/cilium/commit/1af3267d293d93a3aae82d209cb01c015da9015f. As far as I understand, the option was introduced due to complexity issues. Adding some details about that in the commit message, would help people digging through the code in the future.

Similarly to my other comment, I wouldn't make any changes now. Let's wait for the other reviews and take it from there.

sahid commented on bpf/lib/nat.h:
```diff
@@ -1363,6 +1377,7 @@ snat_v6_process(struct __ctx_buff *ctx, enum nat_dir dir,
 	 * add too many unrelated changes.
 	 */
 
+#ifndef DISABLE_SNAT_ICMP_SUPPORT
```
I have compiled out this function. It's not needed if the support of ICMP is disabled. especially with nodeport.

sahid committed:
bpf/nat: implement support of ICMP4 Error packet fragmentation needed

Running in masquerading BPF mode.

Assuming that, Pod is sending a packet to the world of a size bigger
than the accepted MTU in the networking path.

A networking equipment in that networking path would return an ICMP
Error packet with code ICMP_FRAG_NEEDED. According to the RFE 5508,
that networking equipment should embed the original packet in its
response.

Previously because of the missing support of ICMP_FRAG_NEEDED, the NAT
session to return the packet to Pod was never resolved. Meaning that
the ICMP Error packet got dropped.

a) The src IP of the packet becomes the networking equipment that is
generating the error response.  b) The protocol used for the response
is ICMP where the packet sent may be TDP/UDP.

This change is updating the NAT process to read the embedded packet in
the ICMP Error packet that to determine the NAT session and send back
the packet to Pod.

The embbedded packet should also be rev-NATed back so the source IP is
going back to the endpoint IP.

This change is for IPv4.

Fixes: #12968
Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

sahid commented on bpf/lib/nat.h:
```diff
@@ -1168,7 +1427,7 @@ snat_v6_nat(struct __ctx_buff *ctx, const struct ipv6_nat_target *target)
 		/* Letting neighbor solicitation / advertisement pass through. */
 		if (icmp6hdr.icmp6_type == ICMP6_NS_MSG_TYPE ||
-			icmp6hdr.icmp6_type == ICMP6_NA_MSG_TYPE)
+		    icmp6hdr.icmp6_type == ICMP6_NA_MSG_TYPE)
```
Hum will wait for some reviews before to fix this unnecessary move...

joestringer commented on bpf/lib/nat.h:
```diff
@@ -445,18 +534,25 @@ static __always_inline int snat_v4_rewrite_ingress(struct __ctx_buff *ctx,
+			if (ctx_load_bytes(ctx, off +
+					   offsetof(struct icmphdr, type),
+					   &type, 1) < 0)
 				return DROP_WRITE_ERROR;
```
This code now does a `ctx_load_bytes()`. The error should probably not be `DROP_WRITE_ERROR` any more for this case.

joestringer replied:
It looks like `DROP_INVALID` might be the more appropriate code for this case.

joestringer commented on bpf/lib/nat.h:
```diff
@@ -445,18 +534,25 @@ static __always_inline int snat_v4_rewrite_ingress(struct __ctx_buff *ctx,
-			sum_l4 = csum_diff(&from, 4, &to, 4, 0);
-			csum.offset = offsetof(struct icmphdr, checksum);
+			if (type == ICMP_ECHO || type == ICMP_ECHOREPLY) {
+				if (ctx_store_bytes(ctx, off +
```
nit: indentation.

joestringer replied:
^ This probably suggests that your editor is not configured for Linux kernel tab style (tab = 8 width). Depending on your editor, you may be able to find specific instructions about how to configure your editor to format this in the expected style, or alternatively maybe there's a way to use `checkpatch.pl` to format these? (I just have static vim config from long ago that forces my edits to follow the expected indentation, but I appreciate that not everyone has such configuration).

joestringer replied:
I noticed a few other cases like this, including the ICMPv6 case so I won't repeat the comment for other cases. I recognize that you've also noticed some cases yourself. I think that `git show` will format these in the way the kernel expects, so it should be apparent if the indentation isn't aligned quite right.

sahid replied:
Yes right, I'm using emacs, I just fixed it to use `linux` mode and tab 8 width. Thank you.

```
(setq-default c-basic-offset 8 c-default-style "linux")
(setq-default tab-width 8 indent-tabs-mode t)
(define-key c-mode-base-map (kbd "RET") 'newline-and-indent)
```

joestringer commented on bpf/lib/nat.h:
```diff
@@ -1120,6 +1294,83 @@ snat_v6_rev_nat_handle_mapping(struct __ctx_buff *ctx,
 		       NAT_PUNT_TO_STACK : DROP_NAT_NO_MAPPING;
 }
 
+static __always_inline int snat_v6_icmp_rewrite_embedded(struct __ctx_buff *ctx,
```
The commit message here says:

     Tests have been added in a way that the BPF_PROG will act as a
     networking equipment sending the ICMP Error packet.

Where are these tests? It looks like only datapath code changes are included in this PR.

sahid replied:
I had to remove all of them after the switch:
- https://github.com/cilium/cilium/pull/20017

There was initially all written:
- https://github.com/cilium/cilium/commit/3ae101d4531d9e6a8ead75a02b5dd575e27ef472#diff-d41b8c4660415b4abd32bef458ec1bc318b43bb928f8f20027096e4c1abe16dd

I will fix it.

sahid replied:
You can find tests for ipv4. For ipv6 I'm continuing working on them, it's bit more challenging for some reasons.

sahid replied:
@joestringer Hello, what is you feeling regarding tests? For IPv6 enabling `ENABLE_IPV6` is creating verifier error.

```
--- a/bpf/tests/bpf_nat_tests.c                                                                           
+++ b/bpf/tests/bpf_nat_tests.c                                                                           
@@ -7,6 +7,7 @@                                                                                           
 #include <bpf/api.h>                                                                                     
                                                                                                          
 #define ENABLE_IPV4                                                                                      
+#define ENABLE_IPV6                                                                                      
 #define ENABLE_NODEPORT                                                                                  
 #include <node_config.h> 

new coll: program tail_icmp6_handle_ns: load program: permission denied: 36: (67) r1 
<<= 32: R1 pointer arithmetic with <<= operator prohibited (50 line(s) omitted)  
 ```
 
 I was not able to find the problem. Perhaps should I open an issue for that? It may be good if at some point we pass time investigating the reason to then be able writing tests for ipv6, or do you know what I may have missed?

Besides that I think I have addressed all your questions I hope you will have a moment for a second look. Thanks a lot.

sahid replied:
I have reported opened issue for that https://github.com/cilium/cilium/issues/22419

joestringer commented on bpf/lib/nat.h:
```diff
@@ -353,6 +353,92 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+			/* TODO(sahid): There is design issue with
+			 * csum_l4_offset_and_flags as we have to force setting
+			 * attributes for ICMP. We should consider review this
+			 * part.
```
Can you describe this design issue and the impact?

sahid replied:
I removed the whole NOTE, I don't remember exactly what was my issue.

joestringer commented on bpf/lib/nat.h:
```diff
@@ -835,15 +934,83 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
+				 * switched to endpoint.
+				 *
+				 * TODO(sahid): We could certainly support more ICMP
+				 * Error using this function.
```
nit: I don't think it helps to add this todo into the tree. This PR seems to be targeted at `ICMP_DEST_UNREACH` so let's focus on that first. If you want to add additional ICMP error cases, then by all means let's look at that later but I'd suggest just dropping the todo for now.
```suggestion
```

sahid replied:
Sure.

joestringer commented on bpf/lib/nat.h:
```diff
@@ -835,15 +934,83 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
+				 * to determine corresponding protocol, IP src/dst
+				 * of the packet sent to resolve the NAT session.
+				 */
+				tuple.nexthdr = iphdr.protocol;
```
`tuple` was previously initialized using the outer headers in `snat_v4_init_tuple()`, now this same variable is reused to hold the inner tuple. This means that depending on the `icmphdr.type` the `tuple` could represent either the outer or inner tuple. Could that be a problem? If these tuples are somehow different for some reason, it seems like this could trigger mistranslation of the outer header. Should these be tracked in a separate variable to avoid the inner tuple impacting the outer translation?

sahid replied:
It's a bit tricky, but I'm certainly open to change that part. If I use a different  tuple I will have to add conditions to use one or the other when trying to handle the nat mapping.

I was considering that, what we really want to rev-nat is the embedded headers, the outer headers are not really needed as we don't want to do anything with them.

joestringer commented on bpf/lib/nat.h:
```diff
@@ -353,6 +353,92 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
 		       NAT_PUNT_TO_STACK : DROP_NAT_NO_MAPPING;
 }
 
+static __always_inline int snat_v4_icmp_rewrite_embedded(struct __ctx_buff *ctx,
```
Do you think there could be a way to reuse the same implementation for the core rewrite between `snat_v4_rewrite_ingress()` vs. `snat_v4_icmp_rewrite_embedded()`? It seems like as long as you pass the appropriate tuples into the same function, and pull the context to the appropriate offset, then the core function could be identical? Or am I missing something.

joestringer replied:
Ah maybe the checksumming makes this more difficult. :thinking:

sahid replied:
I was initially trying to reuse one of those ingress/egress functions but they differ in implementation a bit more, we also want to change source addr of inner where for ingress we want to fix the dest.

joestringer reviewed (commented):
Thanks for the submission! For what it's worth, I think we could take the first four commits almost as-is if you proposed them in a dedicated PR (just need to fix style nits). This could avoid you having to rebase or carry those patches over time.

I took a look over the code changes and I have some questions below on some aspects of the proposal. See below for more details. Most prominently it would be nice to have in-tree tests for this functionality to confirm that the translation is really occurring as expected. I also noticed a couple of other areas that I couldn't follow very well - one was where the tuple of the outer header was overwritten with the inner tuple's fields which seemed potentially incorrect for unusual packet inputs. The second was whether this proposal could benefit from more code reuse.

I stopped short of reviewing the IPv6 implementation since some of the questions for the IPv4 version could also apply there.

sahid commented on bpf/tests/bpf_nat_tests.c:
```diff
@@ -0,0 +1,440 @@
+	 * NATed to the correct endpoint.  Also, to be valid, the
+	 * embedded packet should be NATed as-well, meaning that the
+	 * source addr of the original packet will be switched from
+	 * IP_HOST to IP_ENDPOINT.
```
The l4 headers will also have to be translated if needed. I will fix that.

sahid commented on bpf/tests/bpf_nat_tests.c:
```diff
@@ -0,0 +1,440 @@
+			int ret;
+
+			/* As a pre-requist we intruct the NAT table
+			 * to simulate an ingress packet sent by
```
s/ingress/egress

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -871,6 +958,66 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
+						return DROP_INVALID;
+					tuple.sport = l4hdr.dport;
+					tuple.dport = l4hdr.sport;
+					break;
```
```suggestion
					if (l4_load_ports(ctx, icmpoff, &tuple.dport) < 0))
						return DROP_INVALID;
					break;
```

sahid replied:
Done thank you

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -871,6 +958,66 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
+				tuple.daddr = iphdr.saddr;
+
+				icmpoff += ipv4_hdrlen(&iphdr);
+				switch (tuple.nexthdr) {
```
Do we also need support for `IPPROTO_SCTP` now?

sahid replied:
If you don't mind, I have opened case #23484, could you assign myself on it? I will work on this missing piece of support but would like if possible not mix more change on this one that to avoid any unpredictable yet blockers, sounds reasonable?

julianwiedmann replied:
Have a look how `snat_v4_rev_nat()` does the port extraction, and `snat_v4_rewrite_ingress()` handles the csum update. Imho it's trivial, and good practice to keep the code base in-sync. It's all hidden behind `ENABLE_SCTP`, so shouldn't impact complexity by default (if that's your concern).

sahid replied:
Sure, I'm not really familiar with this protocol. It's done.

julianwiedmann replied:
Looks good, ty!

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -871,6 +958,66 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
+							   sizeof(l4hdr)) < 0)
+						return DROP_INVALID;
+					tuple.sport = l4hdr.dport;
+					tuple.dport = l4hdr.sport;
```
I'm surprised we're not considering IP Fragmentation here (ie. what if the inner L4 packet doesn't have a L4 header?). But then we're not handling that for the main revSNAT path either...

sahid replied:
It's good point, but wondering whether we could be in that situation since that would probably mean that the DF bit isn't set, right? no reasons that the networking equipment to respond with a need to fragment.

julianwiedmann replied:
Yep, sounds reasonable. Would be worth a comment imho, at some point we will hopefully audit all the remaining paths for missing Fragmentation support (https://github.com/cilium/cilium/issues/11180). So if we think it's not needed here, let's spell that out.

sahid replied:
Done, please let me know if the phrasing sounds right for you :-)

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -372,6 +372,87 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+	if (state->to_dport != tuple->dport) {
+		__be32 suml4 = 0;
+
+		switch (tuple->nexthdr) {
```
SCTP support?

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -372,6 +372,87 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+	 * offset of outer packet.
+	 *  off                     erroff
+	 *   [ ICMP ERROR | INNER IP | INNER L4 ...]
+	 */
```
Let's just make that clear then, and do a `s/off/l4_off` and `s/erroff/inner_l4_off` ?

[I appreciate that you're making them `__u32` btw :+1:]

sahid replied:
Done

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -372,6 +372,87 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+	 *  off                     erroff
+	 *   [ ICMP ERROR | INNER IP | INNER L4 ...]
+	 */
+	erroff = off + sizeof(struct icmphdr) + iphdr.ihl * 4;
```
```suggestion
	erroff = off + sizeof(struct icmphdr) + ipv4_hdrlen(&iphdr);
```

sahid replied:
Done

julianwiedmann reviewed (commented):
Took a peek at the IPv4 code - think the comments so far should also apply for IPv6. Still need to take a deeper look at the CT tuple logic, that stuff is always tricky.

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1157,6 +1308,83 @@ snat_v6_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+	if (ctx_store_bytes(ctx, l4_off + sizeof(struct icmp6hdr) - 4 +
+			    offsetof(struct ipv6hdr, saddr),
+			    &state->to_daddr, 16, 0) < 0)
+		return DROP_WRITE_ERROR;
```
This is `inner_l4_off` and `ipv6_store_saddr()` - right?

sahid replied:
You mean I should use `inner_l4_off` and remove bytes to point on the IP headers? I updated the code to use `ipv6_store_saddr()`.

julianwiedmann replied:
All good - I misread, it would be `inner_l3_off` here of course. Thanks!

sahid committed:
bpf/nat: icmp need-to-frag add support of sctp

This is adding SCTP supports for an ICMP Error need to fragment.

Fixes: #23484,
Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -860,6 +951,70 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
+					tuple.dport = identifier;
+					break;
+				default:
+					return DROP_INVALID;
```
```suggestion
					return DROP_UNKNOWN_L4;
```

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -876,7 +1031,14 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
+			return ret;
+	}
+
+	return snat_v4_rewrite_ingress(ctx, &tuple, state, off, nexthdr);
```
Instead of introducing the `nexthdr` parameter (and adding complexity to `snat_v4_rewrite_ingress()`, why not adjust `tuple->nexthdr` in the `icmp_rewrite_embedded=true` leg so that it reflects the outer header again?

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -372,6 +372,91 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+			break;
+		}
+		default:
+			return DROP_INVALID;
```
```suggestion
			return DROP_UNKNOWN_L4;
```

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1649,6 +1731,72 @@ snat_v6_rev_nat(struct __ctx_buff *ctx, const struct ipv6_nat_target *target)
+			ipv6_addr_copy(&tuple.saddr, (union v6addr *)&iphdr.daddr);
+			ipv6_addr_copy(&tuple.daddr, (union v6addr *)&iphdr.saddr);
+
+			icmpoff += ipv6_hdrlen(ctx, &iphdr.nexthdr);
```
I don't think this will work? `ipv6_hdrlen()` loads from `ctx + ETH_HLEN`, but you want to access the inner headers here. I'd suggest to introduce a `l3_off` parameter.

sahid replied:
I should have fixed it. If you can double check.

julianwiedmann replied:
Please have a closer look.

`ipv6_hdrlen()` needs to walk the packet's Extension headers. But if the inner packet contains **any** relevant extension header (based on `ip6->nexthdr`), then `ipv6_hdrlen()` will currently try to load that extension header from `ctx->data + ETH_LEN + sizeof(struct ipv6hdr)`. So you end up accessing the outer ICMP header instead.

sahid replied:
#24289

sahid replied:
I opened a PR that gives ability to set an offset to `ipv6_hdrlen`, function called `ipv6_hdrlen_from` I hope that makes sense.

julianwiedmann replied:
:+1: Please just include that patch in this PR.

sahid replied:
OK I have cherry-picked the commit in this PR and used the new function. Usage of this new function simplified computing the length :-)

julianwiedmann reviewed (changes requested):
Some comments below.

Two wider concerns:
1. using the same CT tuple and SNAT entry for processing the outer and inner packet feels fragile. I take it there's no way to avoid this?
2. from looking at `snat_v4_rev_nat_handle_mapping()` and `snat_v4_track_connection()`, I believe we can end up double-accounting the inner packet. Looking up the SNAT entry is one thing, but we shouldn't end up calling `ct_lookup4()` for the inner packet.

sahid committed:
ipv6: introduces ipv6 headers length calculation from offset

The (commit: a065036ff010e31199dedfd2bd9e4768fbaedb2c) simplified
prototype of ipv6_hdrlen() to avoid passing offset which was always
set to ETH_HLEN in code base.  Nowaday with the constant involvement
of supporting new protocols that could embed headers in payloads it
looks reasonable to provide a function to easily compute IPv6 headers
from specified offset.

Signed-off-by: Sahid Orentino Ferdjaoui <sahid.ferdjaoui@industrialdiscipline.com>

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1139,6 +1305,86 @@ snat_v6_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+	int hdrlen;
+	__u32 inner_l4_off;
+
+	hdrlen = ipv6_hdrlen(ctx, &tuple->nexthdr);
```
This also needs to be `ipv6_hdrlen_offset()`.

But I don't understand why it's needed in the first place - the caller just extracted the inner CT tuple, so it *must* know the `inner_l3_off` and `inner_l4_off`. Why not simply pass those through?
Same applies for the corresponding IPv4 code.

sahid replied:
++

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -372,6 +372,91 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+	if (ctx_store_bytes(ctx, l4_off + sizeof(struct icmphdr) + offsetof(struct iphdr, saddr),
+			    &state->to_daddr, 4, 0) < 0)
+		return DROP_WRITE_ERROR;
+	if (l3_csum_replace(ctx, l4_off + sizeof(struct icmphdr) + offsetof(struct iphdr, check),
```
```suggestion
	if (ipv4_csum_update_by_diff(ctx, l4_off + sizeof(struct icmphdr),
```

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1460,13 +1778,26 @@ snat_v6_rev_nat(struct __ctx_buff *ctx, const struct ipv6_nat_target *target)
+		/* To rewrite headers for ingress we need the tuple to
+		 * use original next header.
+		 */
+		tuple.nexthdr = nexthdr;
```
This logic can now move up, into the section where we set `icmp_rewrite_embedded = true`. Same for IPv4.

julianwiedmann reviewed (changes requested):
One fix-up needed - but hopefully the affected logic can just go away completely.

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1442,13 +1743,20 @@ snat_v6_rev_nat(struct __ctx_buff *ctx, const struct ipv6_nat_target *target)
+		 * the outer header.
+		 */
+		tuple.nexthdr = nexthdr;
+	} else {
```
Just do this unconditionally inside the `ICMPV6_PKT_TOOBIG` case, and use a `goto`  to skip right to `snat_v6_rewrite_ingress()` ?

sahid replied:
++

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -360,6 +360,82 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+	__be32 sum;
+	struct iphdr iphdr;
+
+	if (ctx_load_bytes(ctx, l4_off + sizeof(struct icmphdr), &iphdr,
```
This should be unused now, right?

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -1393,6 +1608,7 @@ snat_v6_rev_nat(struct __ctx_buff *ctx, const struct ipv6_nat_target *target)
 		__be16 dport;
 	} l4hdr;
 	__u32 off;
+	__u8 nexthdr;
```
Cosmetic nit - we could now have this inside the ICMPV6_PKT_TOOBIG case. Maybe worth cleaning up in a follow-on PR?

sahid replied:
Yes I will do that once merged! Thank you for your great reviews on this change.

sahid replied:
#24537

julianwiedmann reviewed (approved):
Lgtm now, thank you @sahid ! Just one cosmetic nit below.

FWIW, I'm still not fully sold on the re-use of the CT tuple struct. But now that we boiled the change down this much and are no longer touching the main path, I think it's perfectly fine to :ship:.

liuyuan10 commented on bpf/lib/nat.h:
```diff
@@ -360,6 +360,77 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+		/* By recomputing L4 checksum of inner packet we avoid having
+		 * to recompute L4 of the ICMP Error.
+		 */
+		suml4 = csum_diff(&tuple->dport, 4, &state->to_dport, 4, 0);
```
quick question, in this function, where does it apply the L3 pseudo-header csum diff into TCP/UDP header? snat_v4_rewrite_ingress() calls csum_l4_replace() at the end to apply that but this function removed it.

@sahid

sahid replied:
Yes it's good point, looks like we are missing recomputing L4 csum to includes the L3 pseudo-header change on source address.

I have to check that as I have done some tests and tcpdump reported valid csum.

Thank you Yuan.
